### PR TITLE
rebase: use libopenstorage/secrets API to access Vault

### PR DESCRIPTION
**Pull request by nixpanic**
# Describe what this PR does #

This replaces the manual written API access to Vault with the API provided by libopenstorage/secrets. Rook uses the same package to communicate with Vault. Using a maintained package reduces the maintenance burden on Ceph-CSI.

## Is there anything that requires special attention ##

The current feature that uses Vault for secrets should ideally not be broken. Users that consume the current state of the support should transparently be moved to the usage of the new API. This means that the old configuration must remain valid.

## Related issues ##

Fixes: #1645 

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

### CODE REVIEW

**Comment by nixpanic** [comment]
This PR is marked with `[DNM]` and in **draft** state. There are a few `TODO` and `FIXME` notes that need to be addressed before this is complete.

/cc @leseb

File: internal/util/vault.go
Comment (Madhu-1) [inline internal/util/vault.go:214]: When `filepath.Join(kms.vaultPassphrasePath, key)` file will be deleted?

**Review by Madhu-1 — COMMENTED** [review]
can we also have an E2E with vault token-based auth and with certificates later(in next PR's)

**Review by leseb — COMMENTED** [review]
Looks pretty clean so far :) thanks!

File: internal/util/vault.go
Comment (nixpanic) [reply internal/util/vault.go:27]: Sure, can do, but it becomes `loss` :-D

File: internal/util/vault.go
Comment (nixpanic) [reply internal/util/vault.go:214]: Unsure what you mean? This deletes the secret that is identified with the `filepath.Join(kms.vaultPassphrasePath, key)`. It is just the API replacement of doing the manual HTTP DELETE call.

**Comment by nixpanic** [comment]
Oh no, build errors due to updating external-snapshotter in `vendor/`. This update was not intentional. For some reason local builds (and on manual test system) do not have this problem :thinking:

File: internal/util/vault.go
Comment (nixpanic) [reply internal/util/vault.go]: This was modified to `loss`

File: internal/util/crypto.go
Comment (nixpanic) [reply internal/util/crypto.go]: This change has been dropped.

File: internal/util/vault.go
Comment (nixpanic) [reply internal/util/vault.go:218]: Ok, it seems that the secret gets deleted, but is not destroyed by Vault (it becomes empty). Not sure when the deleted secrets get completely removed though. For this, I have updated the e2e test that checks the contents of the secret.

**Comment by nixpanic** [comment]
This is pretty much complete now. Dropping `[DNM]` and **draft** state. More serious reviews welcome :sparkles:

**Comment by nixpanic** [comment]
Many CI jobs fail while deploying rook (maybe when pulling the image from Docker Hub). When #1712 is merged, this PR can be rebased and hopefully the container image gets pulled from the local CI registry instead.

File: internal/util/vault.go
Comment (Madhu-1) [inline internal/util/vault.go:77]: `%s/si/is`

File: internal/util/vault.go
Comment (Madhu-1) [inline internal/util/vault.go:139]: changes not made in this PR can we use `%w` for err instead of `%s`?

File: internal/util/vault.go
Comment (Madhu-1) [inline internal/util/vault.go:158]: FixMe is still needed?

File: internal/util/vault.go
Comment (Madhu-1) [inline internal/util/vault.go:204]: TODO as an enhancement define `passphrase` as a string variable and use it in all functions?

File: internal/util/vault.go
Comment (Madhu-1) [inline internal/util/vault.go:208]: just return an error? or do logging? currently, only error check is done

File: internal/util/vault_test.go
Comment (Madhu-1) [inline internal/util/vault_test.go:2]: 2020?

File: internal/util/vault.go
Comment (Madhu-1) [inline internal/util/vault.go:273]: tmpefile to tmpfile

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
changes look good, small Nit

**Comment by Madhu-1** [comment]
@nixpanic can you please check does these changes cause any backward compatibility issues?

File: internal/util/vault.go
Comment (nixpanic) [reply internal/util/vault.go:77]: ok!

File: internal/util/vault.go
Comment (nixpanic) [reply internal/util/vault.go:139]: sure, will do

File: internal/util/vault.go
Comment (nixpanic) [reply internal/util/vault.go:158]: Well, 'need' maybe not, but it feels like the configuration is done twice, one option might be sufficient. Will keep it as is, it works for now (any likely needs updating when we want to support multiple Vault tokens instead of the ServiceAccount).

File: internal/util/vault.go
Comment (nixpanic) [reply internal/util/vault.go:204]: It is only used 2x, the same as `"data"` (for get/save). I do not think it makes things easier or clearer when there is a variable for these.

File: internal/util/vault.go
Comment (nixpanic) [reply internal/util/vault.go:208]: oh, yes, just returning is sufficient, I think

File: internal/util/vault_test.go
Comment (nixpanic) [reply internal/util/vault_test.go:2]: done

File: internal/util/vault.go
Comment (nixpanic) [reply internal/util/vault.go:273]: done

**Comment by nixpanic** [comment]
> @nixpanic can you please check does these changes cause any backward compatibility issues?

I do not expect any issues. The configuration options are kept the same, and should get converted to the values the API needs. During development, I have swapped out the passphrase fetching/storing with the original implementation, and with that, things keep working (and after image update/downgrade).

**Comment by nixpanic** [comment]
Hmm, deploying Rook fails often now... #1716 might help in troubleshooting the issues

**Comment by nixpanic** [comment]
@mergifyio rebase

**Comment by nixpanic** [comment]
@mergifyio rebase

**Review by Madhu-1 — COMMENTED** [review]
LGTM just waiting for CI to pass

**Comment by nixpanic** [comment]
@Mergifyio rebase

**Comment by nixpanic** [comment]
All of the `/retest` commands were needed because Rook failed to deploy. Logs are non-conclusive, #1719 has been created to gather more logs. Not all tests fail always, so `/retest` may be sufficient and get the approvals for merging. Work on making the CI more stable is ongoing in any case.

**Comment by nixpanic** [comment]
@mergifyio rebase

**Comment by nixpanic** [comment]
@Madhu-1 @leseb all changes have been done, and CI now passed (no Rook deployment failures....)

**Review by Madhu-1 — APPROVED** [review]

**Review by Madhu-1 — APPROVED** [review]

**Comment by tareksha** [comment]
hi @nixpanic this change caused ceph-csi to work with KV-v2 secret mounts only :(
https://github.com/libopenstorage/secrets/blob/a1d19aa9713a820d42fb3743ab8563d192d81ff9/vault/vault.go#L115

**Comment by leseb** [comment]
> hi @nixpanic this change caused ceph-csi to work with KV-v2 secret mounts only :( https://github.com/libopenstorage/secrets/blob/a1d19aa9713a820d42fb3743ab8563d192d81ff9/vault/vault.go#L115

The link tells about the secret backend (kv, transit, etc) not the backend version (v1 or v2). So v1 and v2 are supported.

**Comment by tareksha** [comment]
@leseb Why does ceph-csi care about the details of the underlying secrets engine? it simply does the standard read, write and delete operations against vault. I would like to use a custom plugin, not necessarily KV.
libopenstroage has an as-is support for KV-v1 and KV-v2 only.

**Comment by leseb** [comment]
> @leseb Why does ceph-csi care about the details of the underlying secrets engine? it simply does the standard read, write and delete operations against vault. I would like to use a custom plugin, not necessarily KV. libopenstroage has an as-is support for KV-v1 and KV-v2 only.

Not ceph-csi, the libopenstroage does. It does because it needs to know which API path to query when interacting with the KV Secret Engine. For more details, you can see the [API description](https://www.vaultproject.io/api-docs/secret/kv).
I don't think ceph-csi supports any other Secret Engine than KV when Vault is used. But I'll let @nixpanic confirm that.

**Comment by tareksha** [comment]
@leseb the consumers are not aware of details like the libraries that are used by the internal implementation (neither should they, i guess). what they experience eventually is that ceph-csi works with KV engine only.

**Comment by leseb** [comment]
> @leseb the consumers are not aware of details like the libraries that are used by the internal implementation (neither should they, i guess). what they experience eventually is that ceph-csi works with KV engine only.

I've never said the opposite but that's how it is today, ceph-csi only supports the Vault kv secret engine (v1 and v2). So I'm not sure what point you are trying to make here... Initially, you mentioned only vault secret engine v2 being supported and I can tell you this is not the case.

**Pull request merged**
